c/tskit/tables.c

```diff
@@ -5388,6 +5381,638 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+{
+    int ret = 0;
+
+    if (max_time == 0) {
+        self->max_time = DBL_MAX;
+    } else {
+        self->max_time = max_time;
+    }
+    if (max_time < 0) {
+        ret = TSK_ERR_BAD_PARAM_VALUE;
+    }
+    return ret;
+}
+
+// Define a function that runs the ibd_finder. (Wrapper for the other methods)
+int TSK_WARN_UNUSED
+tsk_ibd_finder_init_and_run(tsk_ibd_finder_t *ibd_finder, tsk_table_collection_t *tables,
+    tsk_id_t *samples, tsk_size_t num_samples, double min_length, double max_time)
+{
+    int ret = 0;
+
+    ret = ibd_finder_init(ibd_finder, samples, num_samples, tables);
+    if (ret != 0) {
+        goto out;
+    }
+
+    ret = tsk_ibd_finder_set_min_length(ibd_finder, min_length);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_ibd_finder_set_max_time(ibd_finder, max_time);
+    if (ret != 0) {
+        goto out;
+    }
+
+    ret = tsk_ibd_finder_run(ibd_finder);
+    if (ret != 0) {
+        goto out;
+    }
+
+out:
+    return ret;
+}
+
+static int TSK_WARN_UNUSED
+ibd_finder_enqueue_segment(
+    tsk_ibd_finder_t *self, double left, double right, tsk_id_t node)
+{
+    int ret = 0;
+    tsk_segment_t *seg;
+    void *p;
+
+    assert(left < right);
+    /* Make sure we always have room for one more segment in the queue so we
+     * can put a tail sentinel on it */
+    if (self->segment_queue_size == self->max_segment_queue_size - 1) {
+        self->max_segment_queue_size *= 2;
+        p = realloc(self->segment_queue,
+            self->max_segment_queue_size * sizeof(*self->segment_queue));
+        if (p == NULL) {
+            ret = TSK_ERR_NO_MEMORY;
+            goto out;
+        }
+        self->segment_queue = p;
+    }
+    seg = self->segment_queue + self->segment_queue_size;
+    seg->left = left;
+    seg->right = right;
+    seg->node = node;
+    self->segment_queue_size++;
+out:
+    return ret;
+}
+
+// static int
+// ibd_finder_find_sample_pair_index(
+//     tsk_ibd_finder_t *self, tsk_id_t sample0, tsk_id_t sample1)
+// {
+//     float pair_index = -1;
+//     size_t index0, index1;
+
+//     // Ensure the sample pairs are in order.
+//     if (sample0 >= sample1) {
+//         goto out;
+//     }
+
+//     index0 = self->sample_id_map[sample0];
+//     index1 = self->sample_id_map[sample1];
+//     pair_index = ((float) self->num_samples) * ((float) self->num_samples - 1) / 2
+//                  - ((float) self->num_samples - (float) index0)
+//                        * ((float) self->num_samples - (float) index0 - 1) / 2
+//                  + (float) index1 - (float) index0 - 1;
+
+//     // Check the result is an integer before converting it.
+//     assert(ceilf(pair_index) == pair_index);
+
+// out:
+//     return (int) pair_index;
+// }
+
+static int
+ibd_finder_find_sample_pair_index2(
+    tsk_ibd_finder_t *self, tsk_id_t sample0, tsk_id_t sample1)
+{
+    int i = 0;
+    while (i < (int) self->num_pairs) {
```

Comment:
Eek! What about:
```c
// change the return type to tsk_id_t so we don't have to cast back and forth to int

tsk_id_t i, s0, s1;

for (i = 0; i < (tsk_id_t) self->num_pairs; i++) {
    s0 = self->pairs[2 * i];
    s1 =self->pairs[2 * i + 1];
    if ((s0 == sample0 && s1 == sample1) || (s0 == sample1 && s1 == sample0)) {
        break;  
    }
    ret = -1;
    if (i < (tsk_id_t) self->num_pairs) {
          ret = tsk_id_t;
    }
    return ret;
} 
```

Reply:
Thanks. I was doing all this int casting deliberately, because I didn't actually realise you could compare for the equality of `tsk_id_t` objects like this! I assumed that for user-defined structs, you'd have to explicitly construct a method to check for equality like `__eq__` in Python. I'm glad that at least some things are easier in C!

I've changed this to be halfway between your suggested function and my original one -- I've kept the return type as an int because it is supposed to be an index, rather than a node ID.  Hopefully it looks a bit neater now though.